libs/user_extensions/src/TTAlpsEvent.cpp

```diff
@@ -125,3 +124,90 @@ bool TTAlpsEvent::ParticleHasISRmotherAfterTopMother(int particleIndex) {
 
   return ParticleHasISRmotherAfterTopMother(motherIndex);
 }
+
+bool TTAlpsEvent::IsGoodMuonFromALP(int muonIndex) {
+
+  auto genParticles = event->GetCollection("GenPart");
+  auto muon = asNanoGenParticle(genParticles->at(muonIndex));
+  
+  if (!muon->IsLastCopy()) return false;
+  if(!muon->IsMuon()) return false;
+
+  int motherIndex = muon->GetMotherIndex();
+  if (motherIndex < 0) return false;
+  auto mother = asNanoGenParticle(genParticles->at(motherIndex));
+
+  // loop over pythia copies of the muon
+  if (mother->IsMuon() == 13) {
+    while(mother->IsMuon() == 13) {
+      motherIndex = muon->GetMotherIndex();
+      if (motherIndex < 0) return false;
+      mother = asNanoGenParticle(genParticles->at(motherIndex));
+    }
+  }
+  // mother must be an ALP
+  if (!mother->IsLastCopy()) return false;
+  if (abs(mother->GetPdgId()) != 54) return false;
+
+  return true;
+}
+
+shared_ptr<PhysicsObjects> TTAlpsEvent::GetGenALPs() {
+  auto genParticles = event->GetCollection("GenPart");
+
+  auto genALPs = make_shared<PhysicsObjects>();
+
+  for (int i = 0; i < genParticles->size(); i++)
+  {
+    auto genParticle = asNanoGenParticle(genParticles->at(i));
+    if(!genParticle->IsGoodParticleWithID(54)) continue;
+    genALPs->push_back(genParticles->at(i));
+  }
+  return genALPs;
+}
+
+shared_ptr<PhysicsObjects> TTAlpsEvent::GetGenMuonsFromALP() {
+  auto genParticles = event->GetCollection("GenPart");
+
+  auto genMuons = make_shared<PhysicsObjects>();
+
+  for (int muon_idx = 0; muon_idx < genParticles->size(); muon_idx++)
+  {
+    auto genParticle = asNanoGenParticle(genParticles->at(muon_idx));
+    int motherIndex = genParticle->GetMotherIndex();
+    if (motherIndex < 0) continue;
+    auto mother = asNanoGenParticle(genParticles->at(motherIndex));
+    if(!IsGoodMuonFromALP(muon_idx)) continue;
+    genMuons->push_back(genParticles->at(muon_idx));
+  }
+  return genMuons;
+}
+
+shared_ptr<PhysicsObjects> TTAlpsEvent::GetMuonsFromALP(shared_ptr<PhysicsObjects> muonCollection, float maxDeltaR) {
+  auto genMuons = GetGenMuonsFromALP();
+
+  auto muonsFromALP = make_shared<PhysicsObjects>();
+  vector<int> savedMuonIndices;
+
+  for (auto genMuon : *genMuons) {
+    auto genMuonp4 = asNanoGenParticle(genMuon)->GetFourVector(0.105);
```

Comment:
Also here, I know 0.105 means muon mass, but to make it clear just define `float muonMass = 0.105` before the loop and use that.